buku

```diff
@@ -200,208 +200,171 @@ class BukuCrypt:
+            enchash = self._decrypt()
+            # Match hash of generated file with that of original DB file
+            dbhash = BukuCrypt.get_filehash(self.dbfile)
+            if dbhash != enchash:
+                os.remove(self.dbfile)
+                raise RuntimeError('Decryption failed')
+            if self.replace:
+                os.remove(self.encfile)
+        except struct.error as e:
+            with contextlib.suppress(FileNotFoundError):
+                os.remove(self.dbfile)
+            raise RuntimeError('Tainted file') from e
         except Exception as e:
             with contextlib.suppress(FileNotFoundError):
-                os.remove(encfile)
-            LOGERR(e)
-            sys.exit(1)
+                os.remove(self.dbfile)
+            raise RuntimeError(e) from e
 
-    @staticmethod
-    def decrypt_file(iterations, dbfile=None):
-        """Decrypt the bookmarks database file.
+    def _cipher(self, key, iv):
+        return self._Cipher(self._algorithms.AES(key), self._modes.CBC(iv), backend=self._default_backend())
 
-        Parameters
-        ----------
-        iterations : int
-            Number of iterations for key generation.
-        dbfile : str, optional
-            Custom database file path (including filename).
-            The '.enc' suffix must be omitted.
-        """
+    def _key(self, salt):
+        key = ('%s%s' % (self.password, salt.decode('utf-8', 'replace'))).encode('utf-8')
+        for _ in range(self.iterations):
+            key = self._sha256(key).digest()
+        return key
 
-        try:
-            from getpass import getpass
-            from hashlib import sha256
+    def _encrypt(self):
+        # Get SHA256 hash of DB file
+        dbhash = BukuCrypt.get_filehash(self.dbfile)
 
-            from cryptography.hazmat.backends import default_backend
-            from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
-        except ImportError:
-            LOGERR('cryptography lib(s) missing')
-            sys.exit(1)
+        # Generate random 256-bit salt and key
+        salt = os.urandom(BukuCrypt.SALT_SIZE)
+        iv = os.urandom(16)
+        encryptor = self._cipher(self._key(salt), iv).encryptor()
+        filesize = os.path.getsize(self.dbfile)
 
-        if iterations < 1:
-            LOGERR('Decryption failed')
-            sys.exit(1)
+        with open(self.dbfile, 'rb') as infp, open(self.encfile, 'wb') as outfp:
+            outfp.write(struct.pack('<Q', filesize))
+            outfp.write(salt)
+            outfp.write(iv)
 
-        if not dbfile:
-            dbfile = os.path.join(BukuDb.get_default_dbdir(), 'bookmarks.db')
-        else:
-            dbfile = os.path.abspath(dbfile)
-            dbpath, filename = os.path.split(dbfile)
+            # Embed DB file hash in encrypted file
+            outfp.write(dbhash)
 
-        encfile = dbfile + '.enc'
+            while chunk := infp.read(BukuCrypt.CHUNKSIZE):
```

Comment:
[Walrus operator (assignment expression)](https://peps.python.org/pep-0572) was added in Python 3.8 (specifically for this sort of thing, even)